SPI upgrade - per-peripheral mutex and GPIO-based SSEL

kjbracey opened the pull request:
### Description

This commit takes some of the work done on the SPI class from #8445, and refines it, to provide the per-peripheral mutex functionality.

This also implements GPIO-based SSEL, which exposes a new `select()/deselect()` API for users to group transfers, and should work on every platform (unlike the HAL-based SSEL). This requires users to use a new constructor to avoid backwards compatibility issues.

To activate the per-peripheral mutex, the HAL must define `SPI_COUNT` and provide `spi_get_peripheral_name()`.  (In #8445 this is a reworked `spi_get_module`, but the name is changed here to avoid a collision with existing HALs - this commit is designed to work without wider HAL changes).

Fixes: #9149

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@jarlamsa, @c1728p9, @ithinuel

kjbracey committed:
SPI upgrade - per-peripheral mutex and GPIO-based SSEL

This commit takes some of the work done on the SPI class from #8445, and
refines it, to provide the per-peripheral mutex functionality.

This also implements GPIO-based SSEL, which exposes a new
select()/deselect() API for users to group transfers, and should work on
every platform (unlike the HAL-based SSEL). This requires users to use a
new constructor to avoid backwards compatibility issues.

To activate the per-peripheral mutex, the HAL must define SPI_COUNT and
provide spi_get_peripheral_name().  (In #8445 this is a reworked
spi_get_module, but the name is changed here to avoid a collision with
existing HALs - this commit is designed to work without wider HAL
changes).

Fixes: #9149

kjbracey commented:
First draft - just for discussion and @jarlamsa to try out at the minute, really.

ciarmcom commented:
@kjbracey-arm, thank you for your changes.
@c1728p9 @ithinuel @jarlamsa @ARMmbed/mbed-os-core @ARMmbed/mbed-os-hal @ARMmbed/mbed-os-maintainers please review.

jarlamsa commented:
```
[Error] SPI.cpp@239,37: 'struct SingletonPtr<mbed::CircularBuffer<mbed::Transaction<mbed::SPI>, 2ul> >' has no member named 'reset'
[Error] SPI.cpp@271,41: 'struct SingletonPtr<mbed::CircularBuffer<mbed::Transaction<mbed::SPI>, 2ul> >' has no member named 'full'
[Error] SPI.cpp@275,41: 'struct SingletonPtr<mbed::CircularBuffer<mbed::Transaction<mbed::SPI>, 2ul> >' has no member named 'push'
[Error] SPI.cpp@323,41: 'struct SingletonPtr<mbed::CircularBuffer<mbed::Transaction<mbed::SPI>, 2ul> >' has no member named 'pop'
```

You probably wanted to use the member access through pointer operator with the access to the transaction_buffer.

jarlamsa commented on drivers/SPI.cpp:
```diff
@@ -167,7 +236,7 @@ void SPI::abort_transfer()
 {
 #if TRANSACTION_QUEUE_SIZE_SPI
-    _transaction_buffer.reset();
+    _peripheral->transaction_buffer.reset();
```
transaction_buffer->reset

jarlamsa commented on drivers/SPI.cpp:
```diff
@@ -199,12 +268,12 @@ int SPI::queue_transfer(const void *tx_buffer, int tx_length, void *rx_buffer, i
     t.width = bit_width;
     Transaction<SPI> transaction(this, t);
-    if (_transaction_buffer.full()) {
+    if (_peripheral->transaction_buffer.full()) {
```
transaction_buffer->full

jarlamsa commented on drivers/SPI.cpp:
```diff
@@ -199,12 +268,12 @@ int SPI::queue_transfer(const void *tx_buffer, int tx_length, void *rx_buffer, i
         core_util_critical_section_enter();
-        _transaction_buffer.push(transaction);
-        if (!spi_active(&_spi)) {
+        _peripheral->transaction_buffer.push(transaction);
```
transaction_buffer->push

jarlamsa commented on drivers/SPI.cpp:
```diff
@@ -250,7 +320,7 @@ void SPI::start_transaction(transaction_t *data)
 {
     Transaction<SPI> t;
-    if (_transaction_buffer.pop(t)) {
+    if (_peripheral->transaction_buffer.pop(t)) {
```
transaction_buffer->pop

jarlamsa commented:
Otherwise seems to be working without issues on NUCLEO_F429ZI when made changes to stm_spi_api.c and TARGET_STM32F4/TARGET_STM32F429xI/TARGET_NUCLEO_F429ZI/PeripheralNames.h

Do we bring the HAL-changes in with this PR or in a separate PR?

ithinuel commented:
Won't this change collide with the work done on [branch](https://github.com/ARMmbed/mbed-os/tree/feature-hal-spec-spi) feature-hal-spec-spi ?

Wouldn't it be better to keep the status quo until the feature gets actually merged into master ?
(I am asking this with in mind the maintenance cost induced by changes on master when time comes to rebase the feature branch.)

kjbracey commented:
Yes, it collides, and the feature branch will need to be rebased and aligned with this :/

The problem is that this feature is required for 5.12, but your feature branch isn't due to land for 5.12.

The system-wide mutex is a blocker for @jarlamsa's project - SD card accesses break the radio driver.

cmonr commented:
@bulislaw @ARMmbed/mbed-os-maintainers Fyi ^^^

deepikabhavnani commented:
> First draft - just for discussion and @jarlamsa to try out at the minute, really.

Will like to see at least some test code to verify the changes in this PR

kjbracey commented:
I'm not sure what the current SPI test infrastructure consists of. I see #7976 in progress for specific testing, but that's part of the pending feature branch, and it's testing the HAL, not `class SPI`. 

For the new select/deselect functionality, bringing it into use by SD drivers and 802.15.4 radio drivers would get that exercised by their test suites. (I've asked @jarlamsa to look at that - would simplify their code and offset code size increase here).

kjbracey commented:
@ithinuel - new commit reworks the peripheral blocks to make them allocated on-demand from the heap. This optimises RAM usage for the typical "SPI peripherals in use much fewer than those available in hardware" case.

I think that's the least worst option, having considered the "static array of spi peripheral count" and "put 1 peripheral object inside each SPI object" options. The last would have been cunning, and optimal for the typical "1 device per peripheral" case, but I couldn't figure out how to hand over correctly and safely when an SPI object was deleted.

Does eliminate the need for `SPI_COUNT` from the HAL, but increases ROM size. Ends up pulling in mutex deletion code in a simple test image, which the static version avoids. Can't see a cunning way around that. (It comes via SPI's virtual destructor).

kjbracey commented:
Requires fix in #9580.

kjbracey committed:
SPIFBlockDevice.h: include SingletonPtr.h

SPIFBlockDevice was using SingletonPtr without an include,
and only getting it via SPI.h.

Spotted while changing SPI to not use SingletonPtr - now
abandoned, but still this shouldn't have been relying on it.

c1728p9 commented on drivers/SPI.cpp:
```diff
@@ -23,33 +24,107 @@
+    _peripheral = SPI::_lookup_peripheral(name);
+    // Otherwise, make a new one
+    if (!_peripheral) {
+        _peripheral.reset(new Peripheral(name));
```
The extra code size from using dynamic memory here will likely be a problem for bootloaders.

c1728p9 reviewed (changes requested):
Requesting changes since the code size increase of dynamic memory will likely be a problem for bootloaders. @kjbracey-arm could you provide details on how much ROM impact this has if the application is otherwise not using dynamic memory? Also, is this only an optimization or is this required to fix problems someone was having?

CC @bulislaw

donatieng commented on components/storage/blockdevice/COMPONENT_SPIF/SPIFBlockDevice.h:
```diff
@@ -16,6 +16,7 @@
 #ifndef MBED_SPIF_BLOCK_DEVICE_H
 #define MBED_SPIF_BLOCK_DEVICE_H
 
+#include "platform/SingletonPtr.h"
```
Is that required?

kjbracey replied:
Yes, the static array of peripheral management blocks contains `Mutex`es, and if those weren't wrapped in a `SingletonPtr`, the array would be placed into RAM even if an image didn't use SPI.

donatieng reviewed (commented):
Great stuff @kjbracey-arm - would it be possible to run Pelion-ready tests on this PR? I think it will exercise this pretty well (e.g. flash chip + connectivity chip on same SPI bus) and give us confidence that this won't break anything.

bulislaw commented on drivers/SPI.cpp:
```diff
@@ -75,69 +141,78 @@ void SPI::frequency(int hz)
 
 void SPI::set_default_write_value(char data)
 {
+    // this does not actually need to lock the peripheral.
```
So could we remove lock/unlock?

kjbracey replied:
Not necessarily - the lock is also functioning as the "API lock". If the SPI object is itself supposed to be thread-safe (as opposed to just thread safety for multiple SPI objects accessing one bus), then there needs to be synchronisation on its members.  It currently seems that such objects are trying to be locally thread-safe, but maybe that needs more thought.

bulislaw reviewed (approved):
Looks good, but the code freeze is approaching. @kjbracey-arm are there any changes needed? @c1728p9 re-review please

kjbracey committed:
Cope with HALs not defining SPIName

kjbracey committed:
Rename SPI_COUNT to DEVICE_SPI_COUNT

Avoid collision with some HALs that already define SPI_COUNT.

ithinuel commented on drivers/SPI.cpp:
```diff
@@ -75,69 +140,78 @@ void SPI::frequency(int hz)
+void SPI::select()
+{
+    lock();
+    if (_select_count++ == 0) {
```
This is just  a matter of coding style, but for clarity, could this increment be done after the if ?

kjbracey replied:
I think I did that quite deliberately because a couple of days before writing it I saw some existing code containing this:

    core_util_atomic_incr_u32(foo, 1);
    if (foo == 1) {

which is not thread-safe, and needs to be

    if (core_util_atomic_incr_u32(foo, 1) == 1) {

I think if people can't get comfortable with the increment/decrement pattern, it just encourages that sort of mistake. (Although breaking it up doesn't hurt in this case, as we are in a lock).

ithinuel commented on drivers/SPI.cpp:
```diff
@@ -75,69 +140,78 @@ void SPI::frequency(int hz)
+
+void SPI::deselect()
+{
+    if (--_select_count == 0) {
```
Like in the previous commend, could this decrement be done before the if to avoid any confusion ?

ithinuel reviewed (changes requested):
Aside from the two "style" comment, LGTM
